### DISCUSSION TITLE
Nushell job control

### DISCUSSION THREAD

**Discussion by shemeshg** [nushell/nushell#5948] [Q&A] [answered]
What is the correct way of doing this in nushell?

# Job control failed in nushell
## from zsh to bash, bash manages job control correctlly
```
zsh% bash

The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.
bash-3.2$ cat
^Z
[1]+  Stopped                 cat
bash-3.2$
```

Nu dont
```
% nu
nu〉cat                                           07/04/2022 11:46:36 PM
^Z
zsh: suspended  nu
%
```

SIGTERM actually managed by Nushell **correctlly**
```
nu〉cat                                           07/04/2022 11:53:41 PM
^C
nu〉
```

## ampersand not accepted in Nu
```
/Users/osx〉do { cat&}                                    07/04/2022 11:47:47 PM
Error: nu::shell::external_command (link)

  × External command
   ╭─[entry #1:1:1]
 1 │ do { cat&}
   ·      ──┬─
   ·        ╰── can't run executable
   ╰────
  help: No such file or directory (os error 2)

nu〉^cat&                                         07/04/2022 11:48:02 PM
Error: nu::shell::external_command (link)

  × External command
   ╭─[entry #2:1:1]
 1 │ ^cat&
   ·  ──┬─
   ·    ╰── can't run executable
   ╰────
  help: No such file or directory (os error 2)

nu〉^cat &                                        07/04/2022 11:48:12 PM
cat: &: No such file or directory
nu〉cat ^                                         07/04/2022 11:48:14 PM
cat: ^: No such file or directory
nu〉cat&                                          07/04/2022 11:48:17 PM
Error: nu::shell::external_command (link)

  × External command
   ╭─[entry #5:1:1]
 1 │ cat&
   · ──┬─
   ·   ╰── can't run executable
   ╰────
  help: No such file or directory (os error 2)

nu〉
```

Is there a Nu way to do that?

**Comment by fdncred** [comment, marked as answer]
No. Nushell does not support job control yet.

**Reply by shemeshg** [reply]
It actually good thing since I don't like much the POSIX way of job control.

My imaginary perfect shell would probably have 2 Modes, one sync and one async.
- Everything after switching  to async mode would run async.
- coordinating between processes would be over "fifo session named pipes",
- streams already read will be automatically disposed 

Something like this pseudocode
```
set mode async
# async run two commands
curl web_cat_pictures|stream cat_stream
curl web_dog_pictures|stream dog_stream
# join results of cat_stream and dog_stream  and wait, and dispose these temporary streams
await cat_stream dog_stream | cat cat_stream dog_stream
set mode sync
# whatever
```

Thanks for you quick response

**Reply by heyakyra** [reply]
@fdncred Is there a ticket to track job control?

**Reply by fdncred** [reply]
I'm not sure. I figure there is because so many have asked for it.

**Reply by BatmanAoD** [reply]
This issue tracks background process control: https://github.com/nushell/nushell/issues/247

**Comment by fdncred** [comment]
I kind of like the way `pueue` does it https://github.com/Nukesor/pueue, but I'm open to other ways.

**Comment by zoefiri** [comment]
would something like this be considerable? 
https://user-images.githubusercontent.com/5888792/161494772-2abccac4-bb92-4a12-9a69-987e66201719.mp4
(from letoram/cat9)

I'm not really sure if multiplexing jobs like this is within the scope of nu but I think it'd allow for a really smooth workflow.

**Reply by fdncred** [reply]
That's very cool, but also a bit confusing.

**Reply by zoefiri** [reply]
admittedly, the way it's presented here is a bit hard on the eyes but I feel like that's just a UI problem. Generally speaking, I just think displaying bkg job output in a UI element above the prompt is a pretty neat feature.

**Comment by yshui** [comment]
Also, without job control, things like suspending vim with Ctrl-Z won't work either.

**Comment by yshui** [comment]
Job control is the only thing that is stopping me from using nushell full time. So I had to take matter into my own hands.

I made a tool that lets you use job control in nushell, just like you normally will:

https://github.com/yshui/job-security

**Reply by fdncred** [reply]
Nice work! I haven't examined the code yet but does it work cross-platform (MacOS, Linux, Windows)?

**Reply by yshui** [reply]
Sorry, this is Linux only.

**Reply by fdncred** [reply]
Dang, that's too bad. I'd love to embed something like this in nushell.

**Reply by shemeshg** [reply]
- I think `jobs` is reserved word on posix....
- Maybe it should be better to registered as Nu extension so you can pipe to a job, `disown` `join`(Wait for a job)  and `kill`

**Reply by fdncred** [reply]
I'd be willing to use the most primitive system that just:
1. spawns a pipeline/executable to a background thread
2. allows you to fg bg/^Z kill
3. shows a list of bg jobs
ok, maybe that's not primitive 😆 I'm just saying it doesn't have to be fancy.

**Reply by yshui** [reply]
@fdncred my tool does all 3, only on Linux though :sweat_smile: Do you run Windows? I am not sure if it would work on MacOS. maybe not that difficult to get it working, but I don't have a MacOS system.

**Reply by yshui** [reply]
> jobs is reserved word on posix

i am not too worried because i am targeting shells that don't support job control.

**Reply by fdncred** [reply]
I run MacOS, Linux, and Windows. I have to in order to test all this cross-platform stuff.

**Reply by BatmanAoD** [reply]
@fdncred @yshui I'm currently using Mac for work, and my personal machines run Windows and WSL2. `job-security` definitely doesn't work out-of-the-box on Mac, but I would be interested in helping to get it working.

**Reply by fdncred** [reply]
I started pulling some "guts" out of pueue into a standalone program, in order to try and get it working for our needs. It doesn't do anything or work, but it seems that pueue has cracked this nut so why not us it?

I'd be willing to put this do-nothing PR up in my repo if people wanted to help and see if we can make something of it.

**Reply by Banyc** [reply]
I tweaked it a bit and made it to the macOS platform, tho its a little bit hacky.

https://github.com/Banyc/job-security/tree/v0.0.2

**Reply by BatmanAoD** [reply]
...sorry, but I'm rescinding my offer to help; I've stopped using nushell for the time being.

**Reply by fdncred** [reply]
Cool. Now we just need one on Windows!
